SimPEG/potential_fields/gravity/_numba_functions.py

```diff
@@ -0,0 +1,177 @@
+                    - kernels[cell_nodes[k, 5]]
+                    - kernels[cell_nodes[k, 6]]
+                    + kernels[cell_nodes[k, 7]]
+                )
+            )
+
+
+def _sensitivity_gravity(
+    receivers,
+    nodes,
+    sensitivity_matrix,
+    cell_nodes,
+    kernel_func,
+    constant_factor,
+):
+    """
+    Fill the sensitivity matrix
+
+    This function should be used with a `numba.jit` decorator, for example:
+
+    ..code::
+
+        from numba import jit
+
+        jit_sensitivity = jit(nopython=True, parallel=True)(_sensitivity_gravity)
+        jit_sensitivity(
+            receivers, nodes, densities, fields, cell_nodes, kernel_func, const_factor
+        )
+
+    Parameters
+    ----------
+    receivers : (n_receivers, 3) array
+        Array with the locations of the receivers
+    nodes : (n_active_nodes, 3) array
+        Array with the location of the mesh nodes.
+    sensitivity_matrix : (n_receivers, n_active_nodes) array
+        Empty 2d array where the sensitivity matrix elements will be filled.
+        This could be a preallocated empty array or a slice of it.
+    cell_nodes : (n_active_cells, 8) array
+        Array of integers, where each row contains the indices of the nodes for
+        each active cell in the mesh.
+    kernel_func : callable
+        Kernel function that will be evaluated on each node of the mesh. Choose
+        one of the kernel functions in ``choclo.prism``.
+    constant_factor : float
+        Constant factor that will be used to multiply each element of the
+        sensitivity matrix.
+
+    Notes
+    -----
+    The constant factor is applied here to each row of the sensitivity matrix
+    because it's more efficient than doing it afterwards: it would require to
+    index the rows that corresponds to each component.
+    """
+    n_receivers = receivers.shape[0]
+    n_nodes = nodes.shape[0]
+    n_cells = cell_nodes.shape[0]
+    # Evaluate kernel function on each node, for each receiver location
+    for i in prange(n_receivers):
+        # Allocate vector for kernels evaluated on mesh nodes
+        kernels = np.empty(n_nodes)
+        for j in range(n_nodes):
+            dx = nodes[j, 0] - receivers[i, 0]
+            dy = nodes[j, 1] - receivers[i, 1]
+            dz = nodes[j, 2] - receivers[i, 2]
+            distance = np.sqrt(dx**2 + dy**2 + dz**2)
+            kernels[j] = kernel_func(dx, dy, dz, distance)
+        # Compute sensitivity matrix elements from the kernel values
+        for k in range(n_cells):
+            sensitivity_matrix[i, k] = constant_factor * (
+                -kernels[cell_nodes[k, 0]]
+                + kernels[cell_nodes[k, 1]]
+                + kernels[cell_nodes[k, 2]]
+                - kernels[cell_nodes[k, 3]]
+                + kernels[cell_nodes[k, 4]]
+                - kernels[cell_nodes[k, 5]]
+                - kernels[cell_nodes[k, 6]]
+                + kernels[cell_nodes[k, 7]]
+            )
```

Comment:
Is it possible to reduce code duplication here and move this to a sub-function?

Reply:
Yes, it's possible. I just moved the evaluation of the kernels and the sums over the nodes for each cell to two different private functions.